onnxruntime/core/optimizer/common_subexpression_elimination.cc

```diff
@@ -0,0 +1,428 @@
+}
+
+bool EquivalenceClass::operator==(const EquivalenceClass& other) const {
+  if (this == &other) {
+    return true;
+  }
+
+  // Below we compare inputs_ as pointers. This is valid due to how EquivalenceClass'es are constructed:
+  // we'll never have two distinct but equal inputs_ here, so their addresses are effectively their value numbers.
+  return hash_ == other.hash_ && output_index_ == other.output_index_ && discriminator_ == other.discriminator_ &&
+         non_op_value_ == other.non_op_value_ &&
+         op_type_ == other.op_type_ && domain_ == other.domain_ &&
+         inputs_ == other.inputs_ &&
+         SameAttributes(attributes_, other.attributes_);
+}
+
+bool EquivalenceClass::operator!=(const EquivalenceClass& other) const {
+  return !operator==(other);
+}
+
+std::size_t EquivalenceClass::CalculateHash() const {
+  std::size_t hash = 0;
+  UpdateHash(output_index_, hash);
+  UpdateHash(discriminator_, hash);
+  UpdateHash(non_op_value_, hash);
+  UpdateHash(op_type_, hash);
+  UpdateHash(domain_, hash);
+  if (attributes_ != nullptr) {
+    for (const auto& kv : *attributes_) {
+      UpdateHash(kv.first, hash);
+      UpdateHash(kv.second, &GetAttributeHash, hash);
+    }
+  }
+
+  for (const auto& arg : inputs_) {
+    for (const EquivalenceClass* input : arg) {
+      UpdateHash(input, DeepPointerHash{}, hash);
+    }
+  }
+
+  return hash;
+}
+
+// Representative of an equivalence class.
+// node_index and output_index define the node that produced the node_arg.
+// For inputs and constant initializers, output_index == kInvalidOutputIndex.
+struct Representative {
+  const NodeArg* node_arg;
+  NodeIndex node_index;
+  OutputIndex output_index;
+};
+
+struct NodeArgPtrHash {
+  std::size_t operator()(const NodeArg* node_arg) const {
+    return std::hash<const NodeArg*>{}(Normalize(node_arg));
+  }
+};
+
+struct NodeArgPtrEquality {
+  bool operator()(const NodeArg* lhs, const NodeArg* rhs) const {
+    return Normalize(lhs) == Normalize(rhs);
+  }
+};
+
+bool IsNodeSupported(const Node& node) {
+  if (node.Domain() != kOnnxDomain || node.ContainsSubgraph()) {
+    return false;
+  }
+
+  if (kNonDeterministicOps.find(node.OpType()) != kNonDeterministicOps.end()) {
+    return false;
+  }
+
+  return true;
+}
+}
+
+}
+
+namespace std {
+template <>
+struct hash<onnxruntime::EquivalenceClass> {
+  std::size_t operator()(const onnxruntime::EquivalenceClass& val) const noexcept {
+    return val.hash_;
+  }
+};
+}
+
+namespace onnxruntime {
+
+Status CommonSubexpressionElimination::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  // Pool of equivalence classes; unique_ptr to guarantee stable address.
+  std::vector<std::unique_ptr<EquivalenceClass>> unique_equivalence_classes;
+
```

Comment:
nit: could do a reserve for a rough guess of the number of entries you'll end up with. graph.NumberOfNodes() will give the number of nodes as a starting point.